File: Python3/740. Delete and Earn.py
class Solution:
    def deleteAndEarn(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        # count = collections.Counter(nums)
        # prev = None
        # avoid = using = 0
        # for k in sorted(count):
        #     if k - 1 != prev:
        #         avoid, using = max(avoid, using), k * count[k] + max(avoid, using)
        #     else:
        #         avoid, using = max(avoid, using), k * count[k] + avoid
        #     prev = k
        # return max(avoid, using)
        
        count = collections.Counter(nums)
        prev = None
        ignore = take = 0
        for k in sorted(count):
            if k-1 != prev:
                ignore, take = max(ignore, take), k*count[k] + max(ignore, take)
            else:
                ignore, take = max(ignore, take), k*count[k] + ignore
            prev = k
        return max(ignore, take)

File: Golang/740. Delete and Earn.go
import "sort"

func deleteAndEarn(nums []int) int {
	numsDict := make(map[int]int, len(nums))
	for _, num := range nums {
		numsDict[num] += num
	}

	sort.Ints(nums)

	ignore := 0
	take := 0
	prev := -100

	for _, num := range nums {
		if num != prev {
			if num != prev+1 {
				ignore, take = max(ignore, take), numsDict[num]+max(ignore, take)
			} else {
				ignore, take = max(ignore, take), numsDict[num]+ignore
			}
			prev = num
		}
	}

	return max(ignore, take)
}

func max(a, b int) int {
	if a > b {
		return a
	} else {
		return b
	}
}
